src/Collection/ImmArray.php

```diff
@@ -484,4 +484,36 @@ protected function arraySort(callable $cb = null)
         }
         return static::fromArray($ar);
     }
+    
+    /**
+     * Checks if element exists in collection
+     *
+     * @param $element
+     * @param bool $strict use strict comparison?
+     * @return bool
+     */
+    public function has($element, $strict = false)
+    {
+        return \in_array($element, $this->toArray(), strict);
+    }
+    
+    /**
+     * Filter out non-unique elements
+     *
+     * @param bool $strict use strict comparison?
+     * @return ImmArray
+     */
+    public function unique($strict = false)
```

Comment:
We could probably use one of these guys to do the same thing more efficiently: http://php.net/manual/en/class.splobjectstorage.php

We can't do a simply `array_flip` and `array_keys`, because that'll only work if your `ImmArray` is a bunch of strings or integers. We'd need to build two sets, one of string/number and another for the SOS. I haven't tried it yet, but I think we could go with something like:

```
public function unique() {
  $uniqueObs = new SplObjectStorage();
  $uniqueScalars = [];
  foreach ($this as $el) {
    if (is_object($el)) $uniqueObs[$el] = true;
    else if (is_array($el)) $uniqueObs[new ArrayObject($el)] = true;
    else $uniqueScalars[$el] = true;
  }
  $sfa = new SplFixedArray(count($uniqueObs) + count($uniqueScalars));
  $i = 0;
  foreach ($uniqueObs as $el => $val) {
    $sfa[$i] = $el;
    $i++;
  }
  foreach ($uniqueScalars as $el => $val) {
    $sfa[$i] = $el;
    $i++;
  }
}
```
This way we rely on the equality computed in the SPL standard, and don't need to create our own rules.

This approach will change the ordering though, which could be a bit of a surprise.

Actually ignore most of the above :)
We should do something closer to your original, but build the `$unique` sets so we can compare. That way we avoid the `n^2` complexity by using a hashmap, but only iterate once to build the new items in the same order they were originally.

```
foreach ($this as $el) {
  if ($uniqueScalars[$el] || $uniqueObs[$el]) {
    if ($uniqueScalars[$el]) $uniqueScalars[$el] = false;
    elseif ($uniqueObs[$el]) $uniqueObs[$el] = false;
    $sfa[$i] = $el;
    $i++;
  }
```

Reply:
This can't be done this way.

- new ArrayObject($el) !== new ArrayObject($el)
- now every scalar can be an array key, part of them are converted i.e. floats to integers

Reply:
Now it's still pesimistic `O(n^2)` but optimistic `O(n*logn)`.